backends/p4fmt/attach.h

```diff
@@ -0,0 +1,41 @@
+#ifndef BACKENDS_P4FMT_ATTACH_H_
+#define BACKENDS_P4FMT_ATTACH_H_
+
+#include "ir/visitor.h"
+#include "lib/source_file.h"
+
+namespace P4::P4Fmt {
+
+class Attach : public Inspector {
+ public:
+    using NodeId = int;
+    struct Comments {
+        std::vector<const Util::Comment *> prefix;
+        std::vector<const Util::Comment *> suffix;
+    };
+    using CommentsMap = std::unordered_map<NodeId, Comments>;
+    enum class TraversalType { Preorder, Postorder };
+
+    explicit Attach(const std::unordered_map<const Util::Comment *, bool> &processedComments)
+        : processedComments(processedComments){};
+
+    void attachCommentsToNode(const IR::Node *, TraversalType);
+
+    bool preorder(const IR::Node *node) override;
+    void postorder(const IR::Node *node) override;
+
+    void addPrefixComments(NodeId, const Util::Comment *);
+    void addSuffixComments(NodeId, const Util::Comment *);
+    const CommentsMap &getCommentsMap() const;
+
+ private:
+    /// This Hashmap tracks each comment’s attachment status to IR nodes. Initially, all comments
+    /// are set to 'false'.
+    std::unordered_map<const Util::Comment *, bool> processedComments;
```

Comment:
Do you really need `bool` value here? Why can't you simply have `std::unordered_set<Util::Comment *>`?

Reply:
Initially, I deleted the comment from the set once it was attached, but modifying the container while iterating over it didn't seem like a good approach. I could use a separate 'Used Comments' set for this, if that seems better? Open to suggestions.

Reply:
I do not see why you need to modify this set during the iteration. Essentially, your code is:

```c++
    for (auto &[comment, isAttached] : processedComments) {
        // Skip if already attached
        if (isAttached) {
            continue;
        }

        switch (...) {
        case A:
            ...
            isAttached = true;
        case B:
            ...
            isAttached = true;
        default:
            error();
        }
}
```

Why do you need to modify anything?
  - The first `continue` seems to be redundant to me, as you always either attaching everything or bail out
  - You are always attaching the whole set.

So why can't you simply attach everything and then do `processedComments.clear()`?

Reply:
Sorry for not clarifying enough. `processedComments` is a shared list of *all* the comments from the file that may get attached to different nodes. I'm using the marking thing to ensure that the same comment won't get attached to another node later. Wouldn't clearing it remove everything from the container, making it difficult to use them for attaching comments to other nodes in the future ? Please let me know if I'm still missing something. :)

Reply:
@asl any further comments?

Reply:
@fruffy I need to study the logic again, will try to do today

Reply:
@snapdgn Ok, so essentially you're having "working set" and "visited list" and you're fusing all of them in the single map.

It seems you're making the copy of this map, is this intentional? Your top-level `globalCommentsMap` will still have all the things there. Maybe you'd need to move it inside? Or take by reference?

Reply:
Yeah, it was not intentional. Modified to take it by ref now. Thanks